Sprint 11

pavelbelenkow opened the pull request.

pavelbelenkow committed:
Connected ProgressHUD dependency
Added loader to fetch auth token method

pavelbelenkow committed:
Minor ui fixes for web view

pavelbelenkow committed:
Implemented algorithm to prevent the race condition in the fetchAuthToken method

pavelbelenkow committed:
Implemented UIBlockingProgressHUD class to synchronize hiding and showing ProgressHUD with UI blocking

pavelbelenkow committed:
Minor fixes in OAuth2Service

pavelbelenkow committed:
Implemented ProfileService
What has been added:
Profile and ProfileResult models
Profile Service Access Layer
fetchProfile method in the splash screen for early loading of profile data
updateProfileDetails method in ProfileViewController for showing uploaded data from Unsplash account

pavelbelenkow committed:
Implemented ProfileImageService
What has been added:
UserResult and Image models
Profile Image Service Access Layer
fetchProfileImageURL method in the splash screen for early loading user avatar url
updateUserImage method in ProfileViewController for future updating of the user's avatar
observation of receiving avatarURL via NotificationCenter tool

pavelbelenkow committed:
Changed old KVO API and implemented new with generics in Web View

pavelbelenkow committed:
Implemented objectTask method using generics to handle the received data from the Network in URLSession extension

pavelbelenkow committed:
Сhanged the previous fetchDataTask method to new objectTask in OAuth2Service, ProfileService and ProfileImageService services
Minor refactoring in these services

pavelbelenkow committed:
Minor fixes in alerts

pavelbelenkow committed:
Kingfisher is attached to the project and imported into ProfileViewController
Implemented feature of uploading and saving profile avatar

pavelbelenkow committed:
SwiftKeychainWrapper is attached to the project and imported into OAuth2TokenStorage
Changed saving and receiving the authorization token via Keychain

pavelbelenkow committed:
Implemented TabBarController programatically

pavelbelenkow committed:
Set UIWindow programmatically
Implemented SplashViewController programmatically

IBDesignable commented on ImageFeed/SplashViewController.swift:
```diff
@@ -6,28 +6,73 @@
+                UIBlockingProgressHUD.dismiss()
+                profileImageService.fetchProfileImageURL(username: profile.username) { _ in }
+            case .failure(let error):
+                let alert = AlertModel(
```
перед тем как презентить алерт надо не забыть UIBlockingProgressHUD.dismiss(), сделать. Что думаешь?

Чтобы не повторяться можно прямо перед switch result  cltkfnm

pavelbelenkow replied:
Да, добавил дисмисс, спасибо за замечание) Когда проверял на ошибку, 5 минут не мог понять, почему не нажимается "ок" на алерте😅

IBDesignable commented on ImageFeed/SplashViewController.swift:
```diff
@@ -6,28 +6,73 @@
+                    self.fetchProfile(token: token)
+                }
+                
+                AlertPresenter(viewController: self).showAlert(model: alert)
```
`AlertPresenter(viewController: self)` - вот эту часть лучше создать во viewDidLoad и обращаться к ней как параметру

pavelbelenkow replied:
Поменял на свойство здесь и в AuthVC ещё)

IBDesignable commented on ImageFeed/Services/ProfileService.swift:
```diff
@@ -0,0 +1,66 @@
+            return
+        }
+        
+        guard var urlComponents = URLComponents(string: Constants.defaultBaseURL) else {
+            return
+        }
+        urlComponents.path = "/me"
+        
+        guard let url = urlComponents.url else {
+            assertionFailure("Failed to create URL")
+            return
+        }
+        
+        var request = URLRequest(url: url)
+        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
```
Хорошо бы URLRequest собирать в отдельном методе :)

pavelbelenkow replied:
Сделал сборку реквеста через статик метод у URLRequest и пролил на сервисы.

IBDesignable commented on ImageFeed/Services/ProfileService.swift:
```diff
@@ -0,0 +1,66 @@
+        let task = session.objectTask(for: request) { [weak self] (result: Result<ProfileResult, Error>) in
+            switch result {
+            case .success(let body):
+                self?.profile = self?.convertProfileResult(from: body)
+                guard let profile = self?.profile else { return }
+                completion(.success(profile))
```
если здесь weak self разыменовать чуть выше, то не придется потом так мучаться с проверкой на нил у self?.profile, код будет чище

pavelbelenkow replied:
Разыменовал, но поле profile получается всё равно придётся проверять, так как он опциональным задан по учебнику(разыменовал и в сервисе аватарки) + получил ответ от наставников, что вроде дополнительно ничего не требуется ещё фиксить в этом участке.
![image](https://github.com/pavelbelenkow/ImageFeed/assets/121253282/5a6d69e5-ec7d-4f3c-992b-0a50d97d9f79)

IBDesignable replied:
Да, просто разыменование не поможет. convertProfileResult  - может выполнять не опциональный Profile и тут мы делаем так:

```swift
let profile = self.convertProfileResult(from: body)
self.profile = profile 
completion(.success(profile)) 
```

IBDesignable commented on ImageFeed/Services/ProfileImageSevice.swift:
```diff
@@ -0,0 +1,65 @@
+            self?.task = nil
+        }
+        self.task = task
+        task.resume()
```
обрати внимание что objectTask уже делает resume. 

Нужно убрать его либо из objectTask, либо здесь не дублироваться (здесь и далее по коду)

pavelbelenkow replied:
Спасибо за подсказку! Уже поменял на всех сервисах)

IBDesignable commented on ImageFeed/Services/OAuth2TokenStorage.swift:
```diff
@@ -14,14 +15,18 @@ final class OAuth2TokenStorage {
         }
         set {
-            userDefaults.set(newValue, forKey: Keys.bearerToken.rawValue)
+            if let newValue {
+                keychainWrapper.set(newValue, forKey: Keys.bearerToken.rawValue)
+            } else {
+                keychainWrapper.removeObject(forKey: Keys.bearerToken.rawValue)
+            }
```
хорошо!

IBDesignable commented on ImageFeed/Models/ProfileResult.swift:
```diff
@@ -0,0 +1,22 @@
+struct ProfileResult: Decodable {
+    let username: String
+    let firstName: String
+    let lastName: String
```
lastName тоже может приходить опциональным, в документации нет, но можно проверить через настройки профиля на сайте
<img width="751" alt="Снимок экрана 2023-05-09 в 21 07 21" src="https://github.com/pavelbelenkow/ImageFeed/assets/121577115/da681ff6-db6b-4a84-a64c-f3a198df4c59">

pavelbelenkow replied:
Я проверял при регистрации и там просто поле last name было обязательным, поэтому опциональным только био было выставлено. Проверил, как это сделала ты, действительно в настройках профиля можно не задавать его)
![image](https://github.com/pavelbelenkow/ImageFeed/assets/121253282/326ba5d8-a71e-48f9-be7b-9386812a8bdb)

IBDesignable commented on ImageFeed/Models/ProfileResult.swift:
```diff
@@ -0,0 +1,22 @@
+    let lastName: String
+    let bio: String?
+    
+    private enum CodingKeys: String, CodingKey {
+        case username
+        case firstName = "first_name"
+        case lastName = "last_name"
+        case bio
+    }
```
Существуют разные модели нейминга на стороне бекенда – camelCase/snake_case, JsonDecoder в swift умеет подстраиваться под нужные, при этом не придется писать ключи самостоятельно.

подробнее можно изучить в официальной доке JSONDecoder или в [этой заметке](https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable)

pavelbelenkow replied:
Классная заметка! Спасибо большое, код из громоздкого превратился в одну лаконичную строчку:3

IBDesignable commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -38,17 +39,21 @@ extension AuthViewController: WebViewViewControllerDelegate {
     // MARK: - Delegate methods
     
     func webViewViewController(_ vc: WebViewViewController, didAuthenticateWithCode code: String) {
+        UIBlockingProgressHUD.show()
+        
         oauth2Service.fetchAuthToken(code) { [weak self] result in
             guard let self else { return }
+            UIBlockingProgressHUD.dismiss()
+            
```
если ты хочешь оставить фетчинг в этом контролере, а не как предлагается в учебнике, то хорошо бы еще webview-дисмисить.

Я сейчас столкнулась с ошибкой видимо при получении профиля, и программа выглядит зависшей на экране веб-вью.

pavelbelenkow replied:
Сделал дисмиссинг вебвью перед запросом на проверку авторизации, вроде теперь скрывается или при ошибке заново будем попадать на вью авторизации.

pavelbelenkow committed:
Changed the dismissing ProgressHUD blocking until receiving response from service
Changed the showing of the alert via the class attribute

pavelbelenkow committed:
The request assembly for services is placed in a single extension via the URLRequest struct method
Added the 'self' dereference in the closure of receiving a response from the network

pavelbelenkow committed:
Implemented the network request via the oauth2Request method in OAuth2Service
Added the 'self' dereference in the closure of receiving response from the network
Removed unnecessary network request resume in ProfileImageService and other services

pavelbelenkow committed:
Made the lastName field optional
Refactoring decoding from snake case to camel case via the keyDecodingStrategy property

pavelbelenkow committed:
Changed the showing of the alert via the class attribute
Added dismissing the webview before sending an authorization request

IBDesignable commented on ImageFeed/Services/ProfileService.swift:
```diff
@@ -25,40 +25,42 @@ final class ProfileService: ProfileServiceProtocol {
-                guard let profile = self?.profile else { return }
+                profile = convertProfileResult(from: body)
+                guard let profile = profile else { return }
                 completion(.success(profile))
```
стало выглядеть лучше :)

pavelbelenkow replied:
Спасибо, но сделал, как ты указала в этом комментарии) https://github.com/pavelbelenkow/ImageFeed/pull/4#discussion_r1194337907

pavelbelenkow committed:
Minor fixes in Profile and ProfileImage services
